### PR TITLE
Initial omslogs logging driver

### PR DESCRIPTION
**Pull request by heaths**
Created a client for the Azure Operations Management Suite (OMS) and a
logging driver to pass stdout/stderr log messages to OMS.

Resolves issue #34257

Signed-off-by: Heath Stewart <heaths@microsoft.com>

**- What I did**

@darobs, @dtzar, @radu-matei, and I created an Azure OMS client and use that in a Docker logging driver to send docker container messages to OMS.

**- How I did it**

Based somewhat on the design of the splunk driver of batching messages we wrote an OMS driver that would queue messages and send in a batch to OMS. Using OMS documentation we allow for logging driver options and some environment variables (for finer-grained customization) to set the workspace ID (sometimes called the customer ID) and the shared key.

**- How to verify it**

1. Build docker with these changes.
2. Create an OMS workspace in Azure and copy the workspaceID and sharedKey.
3. Start a container using the `omslogs` driver:
```
docker run --log-driver omslogs --log-opt "omslogs-workspaceID=$workspaceID" --log-opt "omslogs-sharedKey=$sharedKey" hello-world
```
4. Check the OMS driver logs (may take a couple of minutes to process and appear).

**- Description for the changelog**

> Added built-in logging driver for Azure Operations Management Suite

**- A picture of a cute animal (not mandatory but encouraged)**

waiting squirrel

### CODE REVIEW

**Comment by thaJeztah** [comment]
CI is failing, needs a gofmt;

```
18:07:57 These files are not properly gofmt'd:
18:07:57  - daemon/logdrivers_windows.go
18:07:57 
18:07:57 Please reformat the above files using "gofmt -s -w" and commit the result.
```

File: daemon/logger/omslogs/omsclient/omsclient.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// OMS Data Collector package adapted from
+// https://github.com/Azure/oms-log-analytics-firehose-nozzle
```
Comment (thaJeztah) [inline daemon/logger/omslogs/omsclient/omsclient.go]: Is there no official client that we can vendor?

Comment (heaths) [inline daemon/logger/omslogs/omsclient/omsclient.go]: Not yet. We have started conversations with the Azure team to take ownership of one and would prefer to vendor it.

As to the `gofmt`, I'll do so when I get back from vacation. Having peeked at a few individual build failures, some seemed systemic or maybe the formatting error just wasn't as obvious in some environments as in others.

Thanks!

**Comment by heaths** [comment]
@radu-matei seems you didn't sign-off some of the commits. Please rebase and do so when you come to those commits.

**Comment by AkihiroSuda** [comment]
ping @radu-matei @heaths

**Comment by heaths** [comment]
I'll rebase/squash and make sure commits are signed, but the erred Jenkins build I'm not sure about. I was also under the impression the project wasn't going to take the OMS logger, but if it's still under consideration (since plugins aren't supported on Windows...yet?) I'm happy to get this going.

**Comment by heaths** [comment]
The PowerPC workflow error seems unrelated, but janky yielded warnings as errors:
```
06:58:41 daemon/logger/omslogs/omslogs_test.go:1::warning: file is not goimported (goimports)
06:58:41 daemon/logger/omslogs/omslogs.go:35:46:warning: unnecessary conversion (unconvert)
06:58:41 daemon/logger/omslogs/omslogs.go:36:46:warning: unnecessary conversion (unconvert)
```

File: daemon/logger/omslogs/omslogs.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+	envDomain                = "OMS_LOGGING_DOMAIN"
+	envTimeout               = "OMS_LOGGING_DRIVER_TIMEOUT"
+	envPostMessagesFrequency = "OMS_LOGGING_DRIVER_POST_MESSAGES_FREQUENCY"
+	envPostMessagesBatchSize = "OMS_LOGGING_DRIVER_POST_MESSAGES_BATCH_SIZE"
+	envBufferSize            = "OMS_LOGGING_DRIVER_BUFFER_SIZE"
+	envStreamChannelSize     = "OMS_LOGGING_DRIVER_CHANNEL_SIZE"
+
+	// Default option values
+	defaultTimeout               = time.Duration(5 * time.Second)
+	defaultPostMessagesFrequency = time.Duration(5 * time.Second)
```
Comment (thaJeztah) [inline daemon/logger/omslogs/omslogs.go]: Looks like these should just be `5 * time.Second` (as `time.Second` is already a duration?)

File: daemon/logger/omslogs/omslogs_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
+package omslogs
+
+import (
+	"testing"
+
+	"bytes"
+	"fmt"
+	"github.com/docker/docker/daemon/logger"
+)
```
Comment (thaJeztah) [inline daemon/logger/omslogs/omslogs_test.go]: To fix the CI failure on this file, change to;

```go
import (
	"bytes"
	"fmt"
	"testing"

	"github.com/docker/docker/daemon/logger"
)
```

Comment (heaths) [reply daemon/logger/omslogs/omslogs_test.go]: Thanks for the help. I've made the corrects and pushed the change.

**Comment by heaths** [comment]
The PowerPC build machine is full:

```
18:40:08 Started by user Gordon
18:40:08 Notifying endpoint with url 'https://leeroy.dockerproject.org/notification/jenkins'
18:40:09 [EnvInject] - Loading node environment variables.
18:40:09 Building remotely on power-11 (powerpc power ppc64le) in workspace /var/jenkins_home/workspace/Docker-PRs-powerpc
18:40:09 Running Prebuild steps
18:40:09 FATAL: Unable to produce a script file
18:40:09 java.io.IOException: No space left on device
```

**Comment by thaJeztah** [comment]
thanks! they cleaned up the nodes; and restarted PowerPC

**Comment by thaJeztah** [comment]
oops, wrong button, sorry for that

**Comment by heaths** [comment]
@thaJeztah, what does this mean (janky failed with it right toward the end)?

```
19:14:31 WARNING: deadline exceeded by linter goimports (try increasing --deadline)
19:14:31 Build step 'Execute shell' marked build as failure
```

Is there some way to test all this before I keep pushing? Seems right now I have to commit everything first before I can have the docker containers pull them for build, and only then doesn't it only pull from the moby/moby repo? Seems backward. What if I updated the dockerfiles and hack\make* scripts to maintain backward compatibility, but if no commit ID was passed and the directory exists and is non-empty (i.e. volume-mounted) just skip pulling and build. What that fly?

**Comment by thaJeztah** [comment]
Hm, looks like something made that node timeout during the test 😞 not sure if it's an issue in the code, or just a node that had issues.

To run the integration/unit tests here's some notes; https://gist.github.com/thaJeztah/5b33ca3eafd0d37dbfa5f41e79a57512

> Seems right now I have to commit everything first before I can have the docker containers pull them for build, and only then doesn't it only pull from the moby/moby repo?

For the validation tests, I think you'll have to commit first for some of the tests (dco-check), and it will pull all vendored files to do the vendor-validation. To run just some tests, you may want to just run (e.g.) the `hack/validate/gometalinter` validation;

```
make BIND_DIR=. shell
hack/validate/gometalinter
```

**Comment by thaJeztah** [comment]
ping @anusha-ragunathan - what would be needed to make this work as a plugin (including on Windows? Wondering if that's something that @heaths can assist with)

**Comment by heaths** [comment]
That would be a way to go, but there's already quite a few loggers in there. I would assume the community and maintainers would want to split them out into plugins as well. So wouldn't these be separate efforts?

**Comment by anusha-ragunathan** [comment]
> ping @anusha-ragunathan - what would be needed to make this work as a plugin (including on Windows? Wondering if that's something that @heaths can assist with)

Extracting inbox loggers out as plugins is the future direction. For example, Splunk and SumoLogic have their logging plugins on the Docker Store. https://store.docker.com/search?type=plugin&category=logging. To get started, PTAL at https://docs.docker.com/engine/extend/plugins_logging/ for reference. Windows plugins are not possible at the moment.

**Comment by anusha-ragunathan** [comment]
> Is there, then, a timeline for when plugins on Windows will be supported? 
@johnstep , do you know about the timeline? 

> Will there be a better model to marshal data than across local HTTP?
There are thoughts of using gRPC.

**Comment by johnstep** [comment]
No timeline for Windows managed plugins yet, but we have a design in mind.

**Comment by cpuguy83** [comment]
@heaths The log plugin design is to only use HTTP for control. Log data is streamed over (in Linux) a fifo and encoded as protobuf.

I really don't think we want to bring in another in-tree driver here. It's unfortunate that the Windows side is taking some extra time.

**Comment by heaths** [comment]
@cpuguy83, I can appreciate not cramming value-add into the core code base.

@johnstep, is that a design that could be shared and the community can contribute? I'm happy to convert this to a plugin and contrib what I can to make that possible. But I assume Store changes are necessary for `docker plugin install` to pull Windows plugins as well.

**Comment by johnstep** [comment]
@heaths The two primary issues are (1) packaging containers, since it is infeasible (for multiple reasons) to redistribute an entire Windows root file system, and (2) running containers in privileged containers, as Windows does not support privileged containers.

To address (2), we are considering running plugins directly on the host, which means (1) can be solved by packaging only the plugin files (along with any dependencies not built in to Windows). It's also possible to distribute Windows plugins as layered images, like other Docker images. It might be possible for some plugins to still run in containers, and layered images would allow that.

**Comment by heaths** [comment]
FWIW, I'm starting to look into golang/go#19282 to at least see if we can support a plugin on a Windows host.

**Comment by thaJeztah** [comment]
Let's close this PR for now, because it doesn't look like it will get merged, but feel free to continue the discussion here

**Pull request closed without merging**